src/CBot/CBotInstr/CBotListArray.cpp

```diff
@@ -45,57 +45,111 @@ CBotListArray::~CBotListArray()
 
+    pStk->SetStartError(p->GetStart());
+
     if (IsOfType( p, ID_OPBLK ))
     {
         // each element takes the one after the other
         if (type.Eq( CBotTypArrayPointer ))
         {
-            type = type.GetTypElem();
-
             pStk->SetStartError(p->GetStart());
-            if (nullptr == ( inst->m_expr = CBotListArray::Compile( p, pStk, type ) ))
+            if (p->GetType() == TokenTypVar)
             {
-                goto error;
+                if (!classItem)
+                {
+                    inst->m_expr = CBotTwoOpExpr::Compile(p, pStk);
+                    if (!pStk->GetTypResult().Compare(type))  // compatible type ?
+                    {
+                        pStk->SetError(CBotErrBadType1, p->GetStart());
+                        goto error;
+                    }
+                }
+                else
+                {
+                    pStk->SetError(CBotErrBadLeft, p->GetPrev());
+                    goto error;
+                }
+            }
+            else
+            {
+                if (nullptr == ( inst->m_expr = CBotListArray::Compile( p, pStk, type.GetTypElem() ) ))
+                {
+                    goto error;
+                }
             }
-
             while (IsOfType( p, ID_COMMA ))                                     // other elements?
             {
                 pStk->SetStartError(p->GetStart());
-
-                CBotInstr* i = CBotListArray::Compile(p, pStk, type);
-                if (nullptr == i)
+                CBotInstr* i = nullptr;
+                if (p->GetType() == TokenTypVar)
                 {
-                    goto error;
+                    if (!classItem)
+                    {
+                        i = CBotTwoOpExpr::Compile(p, pStk);
+                        if (nullptr == i || !pStk->GetTypResult().Compare(type))  // compatible type ?
+                        {
+                            pStk->SetError(CBotErrBadType1, p->GetStart());
+                            goto error;
+                        }
+                    }
+                    else
+                    {
+                        pStk->SetError(CBotErrBadLeft, p->GetPrev());
+                        goto error;
+                    }
+                }
+                else
+                {
+                    i = CBotListArray::Compile(p, pStk, type.GetTypElem());
+                    if (nullptr == i)
+                    {
+                        goto error;
+                    }
                 }
-
                 inst->m_expr->AddNext3(i);
+                if ( p->GetType() == ID_COMMA ) continue;
+                if ( p->GetType() == ID_CLBLK ) break;
+
+                pStk->SetError(CBotErrClosePar, p);
+                goto error;
             }
         }
         else
         {
             pStk->SetStartError(p->GetStart());
+            if (classItem && IsOfType(p, TokenTypVar, ID_NEW)) // don't allow func() or var between "{ }"
```

Comment:
I wasn't sure exactly how that should work.
I think the last example maybe did overflow a stack, just not the CBotStack.
I can remove those checks, but I'm not comfortable 
messing with the expression compilation code.
I agree that's how it should be done.